sw/airborne/firmwares/rotorcraft/stabilization/stabilization_attitude_quat_indi.c

```diff
@@ -0,0 +1,322 @@
+
+  quat_from_earth_cmd_i(&stab_att_sp_quat, cmd, heading);
+}
+
+#define BOUND_CONTROLS(_v, _min, _max) {         \
+_v = _v < _min ? _min : _v > _max ? _max : _v;  \
+}
+
+static void attitude_run_indi(int32_t indi_commands[], struct Int32Quat *att_err)
+{
+  //Calculate required angular acceleration
+  struct FloatRates *body_rate = stateGetBodyRates_f();
+  angular_accel_ref.p = reference_acceleration.err_p * QUAT1_FLOAT_OF_BFP(att_err->qx)
+                      - reference_acceleration.rate_p * body_rate->p;
+  angular_accel_ref.q = reference_acceleration.err_q * QUAT1_FLOAT_OF_BFP(att_err->qy)
+                      - reference_acceleration.rate_q * body_rate->q;
+  angular_accel_ref.r = reference_acceleration.err_r * QUAT1_FLOAT_OF_BFP(att_err->qz)
+                      - reference_acceleration.rate_r * body_rate->r;
+
+  //Incremented in angular acceleration requires increment in control input
+  indi_du.p = 1.0/g1.p * (angular_accel_ref.p - filtered_rate_deriv.p);
+  indi_du.q = 1.0/g1.q * (angular_accel_ref.q - filtered_rate_deriv.q);
+  indi_du.r = 1.0/(g1.r+g2) * (angular_accel_ref.r - filtered_rate_deriv.r + g2*indi_du.r);
+
+  //add the increment to the total control input
+  u_in.p = indi_u.p + indi_du.p;
+  u_in.q = indi_u.q + indi_du.q;
+  u_in.r = indi_u.r + indi_du.r;
+
+  //bound the total control input
+  BOUND_CONTROLS(u_in.p, -4500, 4500);
+  BOUND_CONTROLS(u_in.q, -4500, 4500);
+  BOUND_CONTROLS(u_in.r, -4500, 4500);
+
+  //Propagate input filters
+  //first order actuator dynamics
+  u_act_dyn.p = u_act_dyn.p + STABILIZATION_INDI_ACT_DYN_P*( u_in.p - u_act_dyn.p);
+  u_act_dyn.q = u_act_dyn.q + STABILIZATION_INDI_ACT_DYN_Q*( u_in.q - u_act_dyn.q);
+  u_act_dyn.r = u_act_dyn.r + STABILIZATION_INDI_ACT_DYN_R*( u_in.r - u_act_dyn.r);
+
+  //sensor filter
+  stabilization_indi_second_order_filter(&u_act_dyn, &udotdot, &udot, &indi_u, STABILIZATION_INDI_FILT_OMEGA, STABILIZATION_INDI_FILT_ZETA, STABILIZATION_INDI_FILT_OMEGA_R);
+
+  //Don't increment if thrust is off
+  if(stabilization_cmd[COMMAND_THRUST]<300) {
+    FLOAT_RATES_ZERO(indi_u);
+    FLOAT_RATES_ZERO(indi_du);
+    FLOAT_RATES_ZERO(u_act_dyn);
+    FLOAT_RATES_ZERO(u_in);
+    FLOAT_RATES_ZERO(udot);
+    FLOAT_RATES_ZERO(udotdot);
+  }
+  else {
+#if STABILIZATION_INDI_USE_ADAPTIVE
+    lms_estimation();
+#endif
+  }
+
+  /*  INDI feedback */
+  indi_commands[COMMAND_ROLL] = u_in.p;
+  indi_commands[COMMAND_PITCH] = u_in.q;
+  indi_commands[COMMAND_YAW] = u_in.r;
+}
+
+void stabilization_attitude_run(bool_t enable_integrator) {
+
+  /* Propagate the second order filter on the gyroscopes */
+  struct FloatRates *body_rates = stateGetBodyRates_f();
+  stabilization_indi_second_order_filter(body_rates, &filtered_rate_2deriv, &filtered_rate_deriv, &filtered_rate, STABILIZATION_INDI_FILT_OMEGA, STABILIZATION_INDI_FILT_ZETA, STABILIZATION_INDI_FILT_OMEGA_R);
+
+  /* attitude error                          */
+  struct Int32Quat att_err;
```

Comment:
any particular reason to use the fixed point representation here, while the reset of INDI is in float?